discovery/dns/dns.go

```diff
@@ -169,39 +170,117 @@ func (d *Discovery) refresh(ctx context.Context, name string, ch chan<- []*confi
 	return nil
 }
 
-func lookupAll(name string, qtype uint16, logger log.Logger) (*dns.Msg, error) {
+// Try and get an answer for various permutations of the given name,
+// appending the system-configured search path as necessary.
+//
+// There are three possible outcomes:
+//
+// 1. One of the permutations of the given name is recognised as
+//    "valid" by the DNS, in which case we consider ourselves "done"
+//    and that answer is returned.  Note that, due to the way the DNS
+//    handles "name has resource records, but none of the specified type",
+//    the answer received may have an empty set of results.
+//
+// 2.  All of the permutations of the given name are responded to by one of
+//    the servers in the "nameservers" list with the answer "that name does
+//    not exist" (NXDOMAIN).  In that case, it can be considered
+//    pseudo-authoritative that there are no records for that name.
+//
+// 3.  One or more of the names was responded to by all servers with some
+//    sort of error indication.  In that case, we can't know if, in fact,
+//    there are records for the name or not, so whatever state the
+//    configuration is in, we should keep it that way until we know for
+//    sure (by, presumably, all the names getting answers in the future).
+//
+// Outcome 1 is indicated by a valid response message and no error.  Outcome
+// 2 is a nil response message no error.  Outcome 3 is indicated by an error
+// return.  The error will be generic-looking, because trying to return all
+// the errors returned by the combination of all name permutations and
+// servers is a nightmare.
+//
+func lookupWithSearchPath(name string, qtype uint16, logger log.Logger) (*dns.Msg, error) {
 	conf, err := dns.ClientConfigFromFile(resolvConf)
 	if err != nil {
 		return nil, fmt.Errorf("could not load resolv.conf: %s", err)
 	}
 
-	client := &dns.Client{}
 	response := &dns.Msg{}
+	allResponsesValid := true
+
+	for _, lname := range conf.NameList(name) {
+		response, err = lookupName(lname, qtype, conf, logger)
+
+		if err != nil {
+			// We can't go home yet, because a later name
+			// may give us a valid, successful answer.  However
+			// we can no longer say "this name definitely doesn't
+			// exist", because we did not get that answer for
+			// at least one name.
+			allResponsesValid = false
+		} else if response.Rcode == dns.RcodeSuccess {
+			// Outcome 1: GOLD!
+			return response, nil
+		}
+	}
+
+	if allResponsesValid {
+		// Outcome 2: everyone says NXDOMAIN, that's good enough for me
+		return nil, nil
+	}
+	// Outcome 3: boned.
+	return nil, fmt.Errorf("could not resolve %s: one or more search-path-mangled names returned all errors", name)
```

Comment:
I think we should provide a better error message to the user, imho the current one is very hard to understand. At least something like `could not resolve %q: all servers responded with errors to at least one search domain`, but ideally we would include the last error as well.

Reply:
I've updated the error message as suggested.  Adding the last error message seems... fiddly, given that the relevant error variable is now scoped to the loop.  If it's important to report the error, it's probably best just to pass `logger` right down the stack and log all DNS errors.  Noisy, if you've got an environment with flaky DNS, but far more useful than just the last error, in the event of strange problems.

Reply:
Very noisy though. This will easily produce hundreds of log lines for a single issue. But something for a later change.